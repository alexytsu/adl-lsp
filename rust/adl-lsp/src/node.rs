#![allow(dead_code)]
use tree_sitter::Node;

pub enum NodeKind {
    // Comments and Documentation
    Comment,
    Docstring,

    // Identifiers and Names
    Identifier,
    ScopedName,
    TypeName,

    // Type System
    TypeParameters,
    TypeArguments,
    TypeExpression,
    PrimitiveType,

    // Type Definitions
    TypeDefinition,
    NewtypeDefinition,
    StructDefinition,
    UnionDefinition,
    Field,
    FieldBlock,

    // Module System
    Module,
    ModuleDefinition,
    ModuleBody,
    ImportDeclaration,
    ImportPath,

    // Annotations
    Annotation,
    Annotations,
    AnnotationDeclaration,

    // JSON
    JsonValue,
    JsonNumber,
    JsonString,
    JsonArray,
    JsonObject,
    JsonObjectPair,

    // Error
    Error,
}

impl NodeKind {
    pub fn is_user_defined_name(n: &Node) -> bool {
        Self::is_identifier(n) || Self::is_type_name(n) || Self::is_scoped_name(n)
    }

    pub fn is_definition(n: &Node) -> bool {
        Self::is_import_declaration(n) || Self::is_type_name(n) || Self::is_import_declaration(n)
    }

    /// Check if an identifier is part of a scoped name
    pub fn has_scoped_name_parent(n: &Node) -> bool {
        Self::is_identifier(n) && n.parent().is_some_and(|p| Self::is_scoped_name(&p))
    }

    /// Check if a node represents a valid identifier definition that can be referenced
    /// (i.e., an identifier that is part of a type_name)
    pub fn can_be_referenced(n: &Node) -> bool {
        Self::is_identifier(n) && n.parent().is_some_and(|p| Self::is_type_name(&p))
    }
}

impl NodeKind {
    pub fn from_str(s: &str) -> Self {
        match s {
            // Comments and Documentation
            "comment" => Self::Comment,
            "docstring" => Self::Docstring,

            // Identifiers and Names
            "identifier" => Self::Identifier,
            "scoped_name" => Self::ScopedName,
            "type_name" => Self::TypeName,

            // Type System
            "type_parameters" => Self::TypeParameters,
            "type_arguments" => Self::TypeArguments,
            "type_expression" => Self::TypeExpression,
            "primitive_type" => Self::PrimitiveType,

            // Type Definitions
            "type_definition" => Self::TypeDefinition,
            "newtype_definition" => Self::NewtypeDefinition,
            "struct_definition" => Self::StructDefinition,
            "union_definition" => Self::UnionDefinition,
            "field" => Self::Field,
            "field_block" => Self::FieldBlock,

            // Module System
            "module" => Self::Module,
            "module_definition" => Self::ModuleDefinition,
            "module_body" => Self::ModuleBody,
            "import_declaration" => Self::ImportDeclaration,
            "import_path" => Self::ImportPath,

            // Annotations
            "annotation" => Self::Annotation,
            "annotations" => Self::Annotations,
            "annotation_declaration" => Self::AnnotationDeclaration,

            // JSON
            "json_value" => Self::JsonValue,
            "json_number" => Self::JsonNumber,
            "json_string" => Self::JsonString,
            "json_array" => Self::JsonArray,
            "json_object" => Self::JsonObject,
            "json_object_pair" => Self::JsonObjectPair,

            // Default to Error for unknown kinds
            _ => Self::Error,
        }
    }

    pub fn as_str(&self) -> &str {
        match self {
            // Comments and Documentation
            Self::Comment => "comment",
            Self::Docstring => "docstring",

            // Identifiers and Names
            Self::Identifier => "identifier",
            Self::ScopedName => "scoped_name",

            // Type System
            Self::TypeName => "type_name",
            Self::TypeParameters => "type_parameters",
            Self::TypeExpression => "type_expression",
            Self::TypeArguments => "type_arguments",
            Self::PrimitiveType => "primitive_type",

            // Structs
            Self::StructDefinition => "struct_definition",
            Self::Field => "field",
            Self::FieldBlock => "field_block",

            // Unions
            Self::UnionDefinition => "union_definition",

            // Type Definitions
            Self::TypeDefinition => "type_definition",
            Self::NewtypeDefinition => "newtype_definition",

            // Module System
            Self::Module => "module",
            Self::ModuleDefinition => "module_definition",
            Self::ModuleBody => "module_body",
            Self::ImportDeclaration => "import_declaration",
            Self::ImportPath => "import_path",

            // Annotations
            Self::Annotation => "annotation",
            Self::Annotations => "annotations",
            Self::AnnotationDeclaration => "annotation_declaration",

            // JSON
            Self::JsonValue => "json_value",
            Self::JsonNumber => "json_number",
            Self::JsonString => "json_string",
            Self::JsonArray => "json_array",
            Self::JsonObject => "json_object",
            Self::JsonObjectPair => "json_object_pair",

            // Error
            Self::Error => "ERROR",
        }
    }

    // Comments and Documentation
    pub fn is_comment(n: &Node) -> bool {
        n.kind() == Self::Comment.as_str()
    }

    pub fn is_docstring(n: &Node) -> bool {
        n.kind() == Self::Docstring.as_str()
    }

    // Identifiers and Names
    pub fn is_identifier(n: &Node) -> bool {
        n.kind() == Self::Identifier.as_str()
    }

    pub fn is_module_definition(n: &Node) -> bool {
        n.kind() == Self::ModuleDefinition.as_str()
    }

    pub fn is_scoped_name(n: &Node) -> bool {
        n.kind() == Self::ScopedName.as_str()
    }

    // Type System
    pub fn is_type_name(n: &Node) -> bool {
        n.kind() == Self::TypeName.as_str()
    }

    pub fn is_type_parameters(n: &Node) -> bool {
        n.kind() == Self::TypeParameters.as_str()
    }

    pub fn is_type_expression(n: &Node) -> bool {
        n.kind() == Self::TypeExpression.as_str()
    }

    pub fn is_type_arguments(n: &Node) -> bool {
        n.kind() == Self::TypeArguments.as_str()
    }

    pub fn is_primitive_type(n: &Node) -> bool {
        n.kind() == Self::PrimitiveType.as_str()
    }

    // Structs
    pub fn is_struct_definition(n: &Node) -> bool {
        n.kind() == Self::StructDefinition.as_str()
    }

    pub fn is_field_block(n: &Node) -> bool {
        n.kind() == Self::FieldBlock.as_str()
    }

    pub fn is_field(n: &Node) -> bool {
        n.kind() == Self::Field.as_str()
    }

    // Unions
    pub fn is_union_definition(n: &Node) -> bool {
        n.kind() == Self::UnionDefinition.as_str()
    }

    // Type Definitions
    pub fn is_type_definition(n: &Node) -> bool {
        n.kind() == Self::TypeDefinition.as_str()
    }

    pub fn is_newtype_definition(n: &Node) -> bool {
        n.kind() == Self::NewtypeDefinition.as_str()
    }

    // Module System
    pub fn is_module(n: &Node) -> bool {
        n.kind() == Self::Module.as_str()
    }

    pub fn is_module_body(n: &Node) -> bool {
        n.kind() == Self::ModuleBody.as_str()
    }

    pub fn is_import_declaration(n: &Node) -> bool {
        n.kind() == Self::ImportDeclaration.as_str()
    }

    pub fn is_import_path(n: &Node) -> bool {
        n.kind() == Self::ImportPath.as_str()
    }

    // Annotations
    pub fn is_annotation(n: &Node) -> bool {
        n.kind() == Self::Annotation.as_str()
    }

    pub fn is_annotations(n: &Node) -> bool {
        n.kind() == Self::Annotations.as_str()
    }

    pub fn is_annotation_declaration(n: &Node) -> bool {
        n.kind() == Self::AnnotationDeclaration.as_str()
    }

    // JSON
    pub fn is_json_value(n: &Node) -> bool {
        n.kind() == Self::JsonValue.as_str()
    }

    pub fn is_json_number(n: &Node) -> bool {
        n.kind() == Self::JsonNumber.as_str()
    }

    pub fn is_json_string(n: &Node) -> bool {
        n.kind() == Self::JsonString.as_str()
    }

    pub fn is_json_array(n: &Node) -> bool {
        n.kind() == Self::JsonArray.as_str()
    }

    pub fn is_json_object(n: &Node) -> bool {
        n.kind() == Self::JsonObject.as_str()
    }

    pub fn is_json_object_pair(n: &Node) -> bool {
        n.kind() == Self::JsonObjectPair.as_str()
    }

    // Error
    pub fn is_error(n: &Node) -> bool {
        n.kind() == Self::Error.as_str()
    }
}
