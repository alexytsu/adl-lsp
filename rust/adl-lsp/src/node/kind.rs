#![allow(dead_code)]
//! The full grammar sepcification the of ADL's AST from tree-sitter-adl
use tree_sitter::Node;

/// Typed parsing of node.kind() strings from a tree-sitter parse tree
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum NodeKind {
    // Comments and Documentation
    Comment,
    Docstring,
    DefinitionPreamble,

    // Identifiers and Names
    Identifier,
    ScopedName,
    TypeName,

    // Type System
    TypeParameters,
    TypeArguments,
    TypeExpression,
    PrimitiveType,

    // Type Definitions
    TypeDefinition,
    NewtypeDefinition,
    StructDefinition,
    UnionDefinition,
    Field,
    FieldBlock,

    // Module System
    ModuleDefinition,
    ModuleBody,
    ImportDeclaration,
    ImportPath,

    // Annotations
    Annotation,
    Annotations,
    AnnotationDeclaration,
    FieldReference,

    // JSON
    JsonValue,
    JsonNumber,
    JsonString,
    JsonArray,
    JsonObject,
    JsonObjectPair,

    // Error
    Error,
}

impl NodeKind {
    pub fn is_user_defined_name(n: &Node) -> bool {
        Self::is_identifier(n) || Self::is_type_name(n) || Self::is_scoped_name(n)
    }

    pub fn is_local_definition(n: &Node) -> bool {
        Self::is_type_definition(n)
            || Self::is_newtype_definition(n)
            || Self::is_struct_definition(n)
            || Self::is_union_definition(n)
    }

    pub fn is_definition(n: &Node) -> bool {
        Self::is_import_declaration(n) || Self::is_type_name(n) || Self::is_import_declaration(n)
    }

    /// Check if an identifier is part of a scoped name
    pub fn has_scoped_name_parent(n: &Node) -> bool {
        Self::is_identifier(n) && n.parent().is_some_and(|p| Self::is_scoped_name(&p))
    }

    /// Check if a node represents a valid identifier definition that can be referenced
    /// (i.e., an identifier that is part of a type_name)
    pub fn can_be_referenced(n: &Node) -> bool {
        Self::is_identifier(n) && n.parent().is_some_and(|p| Self::is_type_name(&p))
    }
}

impl NodeKind {
    pub fn from_str(s: &str) -> Self {
        match s {
            // Comments and Documentation
            "comment" => Self::Comment,
            "docstring" => Self::Docstring,
            "definition_preamble" => Self::DefinitionPreamble,

            // Identifiers and Names
            "identifier" => Self::Identifier,
            "scoped_name" => Self::ScopedName,
            "type_name" => Self::TypeName,

            // Type System
            "type_parameters" => Self::TypeParameters,
            "type_arguments" => Self::TypeArguments,
            "type_expression" => Self::TypeExpression,
            "primitive_type" => Self::PrimitiveType,

            // Type Definitions
            "type_definition" => Self::TypeDefinition,
            "newtype_definition" => Self::NewtypeDefinition,
            "struct_definition" => Self::StructDefinition,
            "union_definition" => Self::UnionDefinition,
            "field" => Self::Field,
            "field_block" => Self::FieldBlock,

            // Module System
            "module_definition" => Self::ModuleDefinition,
            "module_body" => Self::ModuleBody,
            "import_declaration" => Self::ImportDeclaration,
            "import_path" => Self::ImportPath,

            // Annotations
            "annotation" => Self::Annotation,
            "annotations" => Self::Annotations,
            "annotation_declaration" => Self::AnnotationDeclaration,
            "field_reference" => Self::FieldReference,

            // JSON
            "json_value" => Self::JsonValue,
            "json_number" => Self::JsonNumber,
            "json_string" => Self::JsonString,
            "json_array" => Self::JsonArray,
            "json_object" => Self::JsonObject,
            "json_object_pair" => Self::JsonObjectPair,

            // Default to Error for unknown kinds
            _ => Self::Error,
        }
    }

    pub fn as_str(&self) -> &str {
        match self {
            // Comments and Documentation
            Self::Comment => "comment",
            Self::Docstring => "docstring",
            Self::DefinitionPreamble => "definition_preamble",

            // Identifiers and Names
            Self::Identifier => "identifier",
            Self::ScopedName => "scoped_name",
            Self::TypeName => "type_name",

            // Type System
            Self::TypeParameters => "type_parameters",
            Self::TypeArguments => "type_arguments",
            Self::TypeExpression => "type_expression",
            Self::PrimitiveType => "primitive_type",

            // Type Definitions
            Self::TypeDefinition => "type_definition",
            Self::NewtypeDefinition => "newtype_definition",
            Self::StructDefinition => "struct_definition",
            Self::UnionDefinition => "union_definition",
            Self::Field => "field",
            Self::FieldBlock => "field_block",

            // Module System
            Self::ModuleDefinition => "module_definition",
            Self::ModuleBody => "module_body",
            Self::ImportDeclaration => "import_declaration",
            Self::ImportPath => "import_path",

            // Annotations
            Self::Annotation => "annotation",
            Self::Annotations => "annotations",
            Self::AnnotationDeclaration => "annotation_declaration",
            Self::FieldReference => "field_reference",

            // JSON
            Self::JsonValue => "json_value",
            Self::JsonNumber => "json_number",
            Self::JsonString => "json_string",
            Self::JsonArray => "json_array",
            Self::JsonObject => "json_object",
            Self::JsonObjectPair => "json_object_pair",

            // Error
            Self::Error => "ERROR",
        }
    }

    // Comments and Documentation
    pub fn is_comment(n: &Node) -> bool {
        n.kind() == Self::Comment.as_str()
    }

    pub fn is_docstring(n: &Node) -> bool {
        n.kind() == Self::Docstring.as_str()
    }

    pub fn is_definition_preamble(n: &Node) -> bool {
        n.kind() == Self::DefinitionPreamble.as_str()
    }

    // Identifiers and Names
    pub fn is_identifier(n: &Node) -> bool {
        n.kind() == Self::Identifier.as_str()
    }

    pub fn is_scoped_name(n: &Node) -> bool {
        n.kind() == Self::ScopedName.as_str()
    }

    pub fn is_type_name(n: &Node) -> bool {
        n.kind() == Self::TypeName.as_str()
    }

    // Type System
    pub fn is_type_parameters(n: &Node) -> bool {
        n.kind() == Self::TypeParameters.as_str()
    }

    pub fn is_type_expression(n: &Node) -> bool {
        n.kind() == Self::TypeExpression.as_str()
    }

    pub fn is_type_arguments(n: &Node) -> bool {
        n.kind() == Self::TypeArguments.as_str()
    }

    pub fn is_primitive_type(n: &Node) -> bool {
        n.kind() == Self::PrimitiveType.as_str()
    }

    // Type Definitions
    pub fn is_type_definition(n: &Node) -> bool {
        n.kind() == Self::TypeDefinition.as_str()
    }

    pub fn is_newtype_definition(n: &Node) -> bool {
        n.kind() == Self::NewtypeDefinition.as_str()
    }

    pub fn is_struct_definition(n: &Node) -> bool {
        n.kind() == Self::StructDefinition.as_str()
    }

    pub fn is_union_definition(n: &Node) -> bool {
        n.kind() == Self::UnionDefinition.as_str()
    }

    pub fn is_field(n: &Node) -> bool {
        n.kind() == Self::Field.as_str()
    }

    pub fn is_field_block(n: &Node) -> bool {
        n.kind() == Self::FieldBlock.as_str()
    }

    // Module System
    pub fn is_module_definition(n: &Node) -> bool {
        n.kind() == Self::ModuleDefinition.as_str()
    }

    pub fn is_module_body(n: &Node) -> bool {
        n.kind() == Self::ModuleBody.as_str()
    }

    pub fn is_import_declaration(n: &Node) -> bool {
        n.kind() == Self::ImportDeclaration.as_str()
    }

    pub fn is_import_path(n: &Node) -> bool {
        n.kind() == Self::ImportPath.as_str()
    }

    // Annotations
    pub fn is_annotation(n: &Node) -> bool {
        n.kind() == Self::Annotation.as_str()
    }

    pub fn is_annotations(n: &Node) -> bool {
        n.kind() == Self::Annotations.as_str()
    }

    pub fn is_annotation_declaration(n: &Node) -> bool {
        n.kind() == Self::AnnotationDeclaration.as_str()
    }

    pub fn is_field_reference(n: &Node) -> bool {
        n.kind() == Self::FieldReference.as_str()
    }

    // JSON
    pub fn is_json_value(n: &Node) -> bool {
        n.kind() == Self::JsonValue.as_str()
    }

    pub fn is_json_number(n: &Node) -> bool {
        n.kind() == Self::JsonNumber.as_str()
    }

    pub fn is_json_string(n: &Node) -> bool {
        n.kind() == Self::JsonString.as_str()
    }

    pub fn is_json_array(n: &Node) -> bool {
        n.kind() == Self::JsonArray.as_str()
    }

    pub fn is_json_object(n: &Node) -> bool {
        n.kind() == Self::JsonObject.as_str()
    }

    pub fn is_json_object_pair(n: &Node) -> bool {
        n.kind() == Self::JsonObjectPair.as_str()
    }

    // Error
    pub fn is_error(n: &Node) -> bool {
        n.kind() == Self::Error.as_str()
    }

    pub fn is_missing(n: &Node) -> bool {
        n.is_missing()
    }
}
