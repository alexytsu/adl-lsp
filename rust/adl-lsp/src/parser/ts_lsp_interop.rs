use async_lsp::lsp_types::{Position, Range};
use tree_sitter::{Point, Range as TsRange};

pub fn ts_to_lsp_position(p: &Point) -> Position {
    Position {
        line: p.row as u32,
        character: p.column as u32,
    }
}

pub fn ts_to_lsp_range(r: &TsRange) -> Range {
    Range {
        start: ts_to_lsp_position(&r.start_point),
        end: ts_to_lsp_position(&r.end_point),
    }
}

#[allow(unused)]
pub fn lsp_to_ts_point(p: &Position) -> Point {
    Point {
        row: p.line as usize,
        column: p.character as usize,
    }
}
