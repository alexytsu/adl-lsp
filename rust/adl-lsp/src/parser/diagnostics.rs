use async_lsp::lsp_types::{Diagnostic, DiagnosticSeverity, Range};
use tracing::debug;

use crate::node::{
    AdlAnnotationDeclaration, AdlField, AdlImportDeclaration, AdlModuleBody, AdlModuleDefinition,
    AdlNewtypeDefinition, AdlStructDefinition, AdlTypeDefinition, AdlUnionDefinition, NodeKind,
};
use crate::parser::tree::Tree;
use crate::parser::ts_lsp_interop::ts_to_lsp_position;

use super::ParsedTree;

impl ParsedTree {
    pub fn collect_diagnostics(&self, content: &str) -> Vec<Diagnostic> {
        if content.trim().is_empty() {
            return vec![Diagnostic {
                severity: Some(DiagnosticSeverity::WARNING),
                message: "empty file".to_string(),
                ..Default::default()
            }];
        }

        let mut diagnostics: Vec<Diagnostic> = Vec::new();
        // first collect parse errors
        self.collect_parse_diagnostics(&mut diagnostics);
        self.collect_parse_diagnostics_missing(&mut diagnostics);

        // then collect custom semantic errors
        if let Some(import_diagnostics) = self.collect_import_diagnostics() {
            diagnostics.extend(import_diagnostics);
        }
        if let Some(missing_semicolon_diagnostics) = self.collect_missing_semicolon_diagnostics() {
            diagnostics.extend(missing_semicolon_diagnostics);
        }

        debug!("collected diagnostics: {:?}", diagnostics);
        diagnostics
    }

    pub fn collect_parse_diagnostics(&self, diagnostics: &mut Vec<Diagnostic>) {
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_error)
                .into_iter()
                .map(|n| {
                    let message = match n.parent() {
                        Some(parent) => format!("syntax error in {}", parent.kind()),
                        None => "syntax error".to_string(),
                    };

                    Diagnostic {
                        range: Range {
                            start: ts_to_lsp_position(&n.start_position()),
                            end: ts_to_lsp_position(&n.end_position()),
                        },
                        severity: Some(DiagnosticSeverity::ERROR),
                        message,
                        ..Default::default()
                    }
                }),
        );
    }

    pub fn collect_parse_diagnostics_missing(&self, diagnostics: &mut Vec<Diagnostic>) {
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_missing)
                .into_iter()
                .map(|n| Diagnostic {
                    range: Range {
                        start: ts_to_lsp_position(&n.start_position()),
                        end: ts_to_lsp_position(&n.end_position()),
                    },
                    severity: Some(DiagnosticSeverity::ERROR),
                    message: "missing token '".to_string() + n.kind() + "'",
                    ..Default::default()
                }),
        );
    }

    pub fn collect_missing_semicolon_diagnostics(&self) -> Option<Vec<Diagnostic>> {
        // TODO(alex): abstract this over all nodes with potentially-missing semicolon
        let mut diagnostics = Vec::new();

        // Helper function to create diagnostic for missing semicolon
        let create_missing_semicolon_diagnostic = |n: tree_sitter::Node| Diagnostic {
            range: Range {
                start: ts_to_lsp_position(&n.start_position()),
                end: ts_to_lsp_position(&n.end_position()),
            },
            severity: Some(DiagnosticSeverity::ERROR),
            message: "missing semicolon".to_string(),
            ..Default::default()
        };

        // Check module definitions
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_module_definition)
                .into_iter()
                .filter(|n| {
                    AdlModuleDefinition::try_new(*n)
                        .map(|t| t.is_missing_semicolon())
                        .unwrap_or(false)
                })
                .map(create_missing_semicolon_diagnostic),
        );

        // Check import declarations
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_import_declaration)
                .into_iter()
                .filter(|n| {
                    AdlImportDeclaration::try_new(*n)
                        .map(|t| t.is_missing_semicolon())
                        .unwrap_or(false)
                })
                .map(create_missing_semicolon_diagnostic),
        );

        // Check type definitions
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_type_definition)
                .into_iter()
                .filter(|n| {
                    AdlTypeDefinition::try_new(*n)
                        .map(|t| t.is_missing_semicolon())
                        .unwrap_or(false)
                })
                .map(create_missing_semicolon_diagnostic),
        );

        // Check newtype definitions
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_newtype_definition)
                .into_iter()
                .filter(|n| {
                    AdlNewtypeDefinition::try_new(*n)
                        .map(|t| t.is_missing_semicolon())
                        .unwrap_or(false)
                })
                .map(create_missing_semicolon_diagnostic),
        );

        // Check struct definitions
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_struct_definition)
                .into_iter()
                .filter(|n| {
                    AdlStructDefinition::try_new(*n)
                        .map(|t| t.is_missing_semicolon())
                        .unwrap_or(false)
                })
                .map(create_missing_semicolon_diagnostic),
        );

        // Check union definitions
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_union_definition)
                .into_iter()
                .filter(|n| {
                    AdlUnionDefinition::try_new(*n)
                        .map(|t| t.is_missing_semicolon())
                        .unwrap_or(false)
                })
                .map(create_missing_semicolon_diagnostic),
        );

        // Check fields
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_field)
                .into_iter()
                .filter(|n| {
                    AdlField::try_new(*n)
                        .map(|t| t.is_missing_semicolon())
                        .unwrap_or(false)
                })
                .map(create_missing_semicolon_diagnostic),
        );

        // Check annotation declarations
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_annotation_declaration)
                .into_iter()
                .filter(|n| {
                    AdlAnnotationDeclaration::try_new(*n)
                        .map(|t| t.is_missing_semicolon())
                        .unwrap_or(false)
                })
                .map(create_missing_semicolon_diagnostic),
        );

        Some(diagnostics)
    }

    pub fn collect_import_diagnostics(&self) -> Option<Vec<Diagnostic>> {
        let imports = self.find_all_nodes(NodeKind::is_import_declaration);

        let module_body = AdlModuleBody::try_new(self.find_first_node(NodeKind::is_module_body)?)?;
        let mut cursor = module_body.cursor();
        cursor.goto_first_child(); // opening module brace

        let mut first_non_import = None;
        while cursor.goto_next_sibling() {
            let node = cursor.node();
            if !NodeKind::is_import_declaration(&node)
                && !NodeKind::is_docstring(&node)
                && !NodeKind::is_comment(&node)
            {
                first_non_import = Some(node);
                break;
            }
        }

        let first_non_import = first_non_import?;

        let out_of_order_imports = imports
            .iter()
            .filter_map(|node| {
                // imports should only be at the top of a module
                if node.start_position() > first_non_import.start_position() {
                    Some(Diagnostic {
                        range: Range {
                            start: ts_to_lsp_position(&node.start_position()),
                            end: ts_to_lsp_position(&node.end_position()),
                        },
                        message: "imports must be declared at the beginning of a module"
                            .to_string(),
                        severity: Some(DiagnosticSeverity::ERROR),
                        ..Default::default()
                    })
                } else {
                    None
                }
            })
            .collect();

        // TODO: attempt to resolve the imports and report errors for invalid imports

        Some(out_of_order_imports)
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::Url;
    use insta::assert_yaml_snapshot;

    use crate::parser::AdlParser;

    #[test]
    fn test_collect_parse_error() {
        let url: Url = "file://foo/error.adl".parse().unwrap();
        let contents = include_str!("input/error.adl");

        let parsed = AdlParser::new().parse(url.clone(), contents);
        assert!(parsed.is_some());
        assert_yaml_snapshot!(parsed.unwrap().collect_diagnostics(contents));
    }

    #[test]
    fn test_collect_import_error() {
        let url: Url = "file://foo/importerror.adl".parse().unwrap();
        let contents = include_str!("input/importerror.adl");

        let parsed = AdlParser::new().parse(url.clone(), contents);
        assert!(parsed.is_some());
        assert_yaml_snapshot!(parsed.unwrap().collect_diagnostics(contents));
    }

    #[test]
    fn test_collect_missing_semicolon_error() {
        let url: Url = "file://foo/missing_semicolons.adl".parse().unwrap();
        let contents = include_str!("input/missing_semicolons.adl");

        let parsed = AdlParser::new().parse(url.clone(), contents);
        assert!(parsed.is_some());
        assert_yaml_snapshot!(parsed.unwrap().collect_diagnostics(contents));
    }
}
