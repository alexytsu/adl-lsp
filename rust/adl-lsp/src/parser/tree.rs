use async_lsp::lsp_types::Position;
use tree_sitter::{Node, TreeCursor};

use crate::parser::ts_lsp_interop as ts_lsp;
use crate::{node::NodeKind, parser::ParsedTree};

/// Basic tree traversal methods for working with the tree-sitter parsed tree
pub trait Tree {
    fn advance_cursor_to(cursor: &mut TreeCursor<'_>, nid: usize) -> bool;
    fn find_all_nodes_from<'a>(&self, n: Node<'a>, f: fn(&Node) -> bool) -> Vec<Node<'a>>;
    fn walk_and_filter<'a>(
        cursor: &mut TreeCursor<'a>,
        f: fn(&Node) -> bool,
        early: bool,
    ) -> Vec<Node<'a>>;
    fn get_node_at_position<'a>(&'a self, pos: &Position) -> Option<Node<'a>>;
    fn find_all_nodes(&self, f: fn(&Node) -> bool) -> Vec<Node>;
    fn find_first_node<'a>(&'a self, f: fn(&Node) -> bool) -> Option<Node<'a>>;
    fn find_node_from<'a>(&self, n: Node<'a>, f: fn(&Node) -> bool) -> Vec<Node<'a>>;
}

impl Tree for ParsedTree {
    fn walk_and_filter<'a>(
        cursor: &mut TreeCursor<'a>,
        f: fn(&Node) -> bool,
        early: bool,
    ) -> Vec<Node<'a>> {
        let mut v = vec![];

        loop {
            let node = cursor.node();

            if f(&node) {
                v.push(node);
                if early {
                    break;
                }
            }

            if cursor.goto_first_child() {
                v.extend(Self::walk_and_filter(cursor, f, early));
                cursor.goto_parent();
            }

            if !cursor.goto_next_sibling() {
                break;
            }
        }

        v
    }

    fn advance_cursor_to(cursor: &mut TreeCursor<'_>, nid: usize) -> bool {
        loop {
            let node = cursor.node();
            if node.id() == nid {
                return true;
            }
            if cursor.goto_first_child() {
                if Self::advance_cursor_to(cursor, nid) {
                    return true;
                }
                cursor.goto_parent();
            }
            if !cursor.goto_next_sibling() {
                return false;
            }
        }
    }

    fn get_node_at_position<'a>(&'a self, pos: &Position) -> Option<Node<'a>> {
        let pos = ts_lsp::lsp_to_ts_point(pos);
        self.tree.root_node().descendant_for_point_range(pos, pos)
    }

    fn find_all_nodes(&self, f: fn(&Node) -> bool) -> Vec<Node> {
        self.find_all_nodes_from(self.tree.root_node(), f)
    }

    fn find_all_nodes_from<'a>(&self, n: Node<'a>, f: fn(&Node) -> bool) -> Vec<Node<'a>> {
        let mut cursor = n.walk();
        Self::walk_and_filter(&mut cursor, f, false)
    }

    fn find_first_node<'a>(&'a self, f: fn(&Node) -> bool) -> Option<Node<'a>> {
        self.find_node_from(self.tree.root_node(), f).first().copied()
    }

    fn find_node_from<'a>(&self, n: Node<'a>, f: fn(&Node) -> bool) -> Vec<Node<'a>> {
        let mut cursor = n.walk();
        Self::walk_and_filter(&mut cursor, f, true)
    }
}

impl ParsedTree {
    pub fn get_identifier_at<'a>(
        &'a self,
        pos: &Position,
        content: &'a [u8],
    ) -> Option<(&'a str, Node<'a>)> {
        self.get_node_at_position(pos)
            .filter(NodeKind::is_identifier)
            .map(|n| (n.utf8_text(content.as_ref()).expect("utf-8 parse error"), n))
    }
}
