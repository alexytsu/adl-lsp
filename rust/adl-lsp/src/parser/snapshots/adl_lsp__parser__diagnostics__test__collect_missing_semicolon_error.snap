---
source: src/parser/diagnostics.rs
expression: parsed.unwrap().collect_diagnostics(contents)
---
- range:
    start:
      line: 18
      character: 42
    end:
      line: 18
      character: 43
  severity: 1
  message: syntax error in annotation_declaration
- range:
    start:
      line: 0
      character: 0
    end:
      line: 19
      character: 1
  severity: 1
  message: missing semicolon
- range:
    start:
      line: 1
      character: 4
    end:
      line: 1
      character: 22
  severity: 1
  message: missing semicolon
- range:
    start:
      line: 2
      character: 4
    end:
      line: 2
      character: 21
  severity: 1
  message: missing semicolon
- range:
    start:
      line: 4
      character: 4
    end:
      line: 4
      character: 24
  severity: 1
  message: missing semicolon
- range:
    start:
      line: 6
      character: 4
    end:
      line: 6
      character: 42
  severity: 1
  message: missing semicolon
- range:
    start:
      line: 8
      character: 4
    end:
      line: 11
      character: 5
  severity: 1
  message: missing semicolon
- range:
    start:
      line: 13
      character: 4
    end:
      line: 16
      character: 5
  severity: 1
  message: missing semicolon
- range:
    start:
      line: 9
      character: 8
    end:
      line: 9
      character: 19
  severity: 1
  message: missing semicolon
- range:
    start:
      line: 10
      character: 8
    end:
      line: 10
      character: 17
  severity: 1
  message: missing semicolon
- range:
    start:
      line: 14
      character: 8
    end:
      line: 14
      character: 19
  severity: 1
  message: missing semicolon
- range:
    start:
      line: 15
      character: 8
    end:
      line: 15
      character: 17
  severity: 1
  message: missing semicolon
- range:
    start:
      line: 18
      character: 4
    end:
      line: 18
      character: 51
  severity: 1
  message: missing semicolon
